### DISCUSSION TITLE
Declaring finite sorts

### DISCUSSION THREAD

**Discussion by nicola-gigante** [cvc5/cvc5#9171] [Q&A] [answered]
Hi,

is there a way to declare an uninterpreted finite sort in `cvc5`, similar to `EnumSort` of the Z3 API? I mean, without declaring a generic sort and then adding constraints and axioms to make it finite.

**Comment by ajreynol** [comment]
We don't currently have a generic Enum type.

An Enum type for us would be syntax-sugar for a datatype with a set of 0-argument constructors, which is currently the best way to specify this.

We could consider adding an API function for Enum types to make the API calls simpler. Is this something you would be interested in?

**Reply by nicola-gigante** [reply]
Hi, thanks for the reply!

I'm working on a tool that supports both Z3 and cvc5 as alternative backends and I'm limited to the features that are in common to both, so I benefit when there is the most similarity between the two APIs. That said, whether adding a new API is worth the effort also depends on the performance implications. Currently I'm emulating the enum sort with something along the lines of the following:

```
(set-logic ALL)

(declare-sort S 0)
(declare-const c1 S)
(declare-const c2 S)
(declare-const c3 S)

(assert (distinct c1 c2 c3))
(assert (forall ((x S)) (not (distinct x c1 c2 c3))))

```

What do you think?
How will this perform in practice w.r.t. Z3 support for Enum types or w.r.t to a set of datatype constructors in cvc5?

**Comment by ajreynol** [comment, marked as answer]
It is significantly better to use datatypes for this:
```
(declare-datatype S ((c1) (c2) (c3)))
```
This avoids the use of quantified formulas.  Above, c1, c2, c3 are distinct constants such that all terms of type S are either c1, c2 or c3.

For details on this theory, see, e.g.:
https://homepage.divms.uiowa.edu/~ajreynol/cade15.pdf
https://homepage.cs.uiowa.edu/~tinelli/papers/BarST-PDPAR-06.pdf

**Reply by nicola-gigante** [reply]
Ah yes, thanks, now I see that datatypes are supported by Z3 as well, so it's the right feature for my use case. Can I ask you to post a snippet to declare the equivalent of the SMTLIB line above with cvc5's C++ API?

**Reply by barrettcw** [reply]
Have a look at the datatypes API documentation which shows how to do the same thing in SMT-LIB and the various APIs:
https://cvc5.github.io/docs/cvc5-1.0.2/examples/datatypes.html

**Reply by nicola-gigante** [reply]
Thanks! That's exactly what I was looking for!
